### PR TITLE
[SPARK-16629] Allow comparisons between UDTs and Datatypes

### PR DESCRIPTION
**Pull request by damnMeddlingKid**
## What changes were proposed in this pull request?

Fixes https://issues.apache.org/jira/browse/SPARK-16629

Currently UDTs can not be compared to Datatypes even if their sqlTypes match. this leads to errors like this 

```

In [12]: filtered = df.filter(df['udt_time'] > threshold)
---------------------------------------------------------------------------
AnalysisException                         Traceback (most recent call last)
/Users/franklyndsouza/dev/starscream/bin/starscream in <module>()
----> 1 thresholded = df.filter(df['udt_time'] > threshold)

AnalysisException: u"cannot resolve '(`udt_time` > TIMESTAMP('2015-10-20 01:00:00.0'))' due to data typ mismatch: '(`udt_time` > TIMESTAMP('2015-10-20 01:00:00.0'))' requires (boolean or tinyint or smallint or int or bigint or float or double or decimal or timestamp or date or string or binary) type, not pythonuserdefined"

```

This PR adds some methods that allow UDTs to be correctly compared to a Datatype.
## How was this patch tested?

Built locally and tested in the pyspark repl.

@angelini @davies @rxin

### CODE REVIEW

**Comment by srowen** [comment]
Please read https://cwiki.apache.org/confluence/display/SPARK/Contributing+to+Spark and attach this to your JIRA via the title. I fixed some fields in your JIRA

**Comment by rxin** [comment]
cc @cloud-fan

**Comment by cloud-fan** [comment]
can you add a regression test in your PR? thanks

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 1]
```diff
@@ -110,6 +110,28 @@ class DataFrameSuite extends QueryTest with SharedSQLContext {
     )
   }
 
+  test("test filtering with predicates on UDT columns") {
+    val rowRDD = sparkContext.parallelize(Seq(Row(new ExampleMoney(1.0)), Row(new ExampleMoney(2.0)), Row(new ExampleMoney(3.0))))
+    val schema = StructType(Array(StructField("dollar", new ExampleMoneyUDT(), false)))
+    val df = spark.createDataFrame(rowRDD, schema)
+
+    checkAnswer(df.filter(df("dollar") < 2.0), Seq(Row(new ExampleMoney(1.0))))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala]: cc @mengxr , is UDT designed to work like this?

**Comment by damnMeddlingKid** [comment]
I've tested this successfully with int and timestamp types, but it doesn't seem to work with DecimalType. Anyone know what could be wrong ?

**Pull request closed without merging**